File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/RotateAdminRootViewModel.kt
// Slice D (docs/device-admin-tier-spec.md §5) — owner-facing "Rotate admin key".
//
// The owner mints a FRESH random admin master root and proves the hand-off with
// an old-root-signs-new-root `AdminRootRotation`. `.com` relays the signed proof;
// each box re-pins to the new root ONLY after verifying the proof against its
// pinned OLD root (never `.com`'s word — the box-side trust anchor stays the
// pinned key, §5). The OLD root reads are biometric-gated (signing a rotation is
// never silent).
//
// REVOKE SEMANTIC (§4.3): rotation EXCLUDES other admin devices that hold only
// the OLD bare root — once boxes re-pin, their old-root signatures stop
// verifying. That is the whole point of a rotation as a "cut off a lost/rogue
// admin" remedy; it is NOT additive. Grant-based admins are dropped via the
// separate grant-revocation path.
//
// ORDERING INVARIANTS: the rotation is COMMITTED (proof posted + new root
// stored device-local) before the recovery re-escrow step ever runs — the
// re-escrow (D-3) is a best-effort FOLLOW-UP that can never fail or unwind the
// rotation. When cloud recovery is enrolled the flow parks in
// [RotateAdminRootPhase.DoneNeedsRecoveryUpdate]: without a re-escrow the
// recovery envelope still wraps the DEAD old root, so a post-rotation
// credential recovery would restore a root no box accepts. The re-escrow is
// interactive (recovery passphrase + WebAuthn PRF assert EMIT the wrap key —
// consent-as-crypto, never a boolean), retryable on failure, and explicitly
// skippable (the UI warns). The NEW seed is held privately by the VM for the
// re-escrow step only — it never rides a Phase data class.

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.AdminRootRotationRequest
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.core.AdminRootRotation
import com.flagshipserver.app.core.AdminRootRotationClaim
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.keystore.Keystore
import com.google.crypto.tink.subtle.Ed25519Sign
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.security.SecureRandom

sealed interface RotateAdminRootPhase {
    data object Idle : RotateAdminRootPhase
    data object Rotating : RotateAdminRootPhase
    /** Success — carries the NEW admin-root pubkey (hex) now pinned locally. */
    data class Done(val newAdminRootPubHex: String) : RotateAdminRootPhase
    /** Rotation SUCCEEDED but cloud recovery still escrows the DEAD old root.
     *  The UI collects the recovery passphrase → [RotateAdminRootViewModel
     *  .updateRecoveryBackup] (retryable via [errorMessage]) or skips. */
    data class DoneNeedsRecoveryUpdate(
        val newAdminRootPubHex: String,
        val errorMessage: String? = null,
        val updating: Boolean = false,
    ) : RotateAdminRootPhase
    data class Failed(val message: String) : RotateAdminRootPhase
}

class RotateAdminRootViewModel(
    private val server: FlagshipServerClient,
    private val username: String,
    /** Mints the NEW random 32-byte admin-root seed. Injectable for
     *  deterministic tests. */
    private val mintSeed: () -> ByteArray = { ByteArray(32).also { SecureRandom().nextBytes(it) } },
    private val now: () -> Long = { System.currentTimeMillis() },
    /** True iff this device holds the admin master root. */
    private val hasAdminRoot: () -> Boolean = { Keystore.hasAdminRoot() },
    /** Loads the OLD admin-root SIGNER (biometric-gated in production). */
    private val loadOldSigner: suspend () -> Ed25519Sign = { Keystore.adminRootKey("Rotate your admin key") },
    /** The OLD admin-root pubkey hex held on this device. */
    private val oldPubHex: () -> String? = { Keystore.adminRootPubHex() },
    /** Re-store the NEW root device-local (replaces the old one). */
    private val storeNewRoot: (ByteArray) -> Unit = { Keystore.importAdminRoot(it) },
    /** Whether cloud recovery is enrolled — gates the post-rotation re-escrow
     *  prompt. Errors ⇒ best-effort skip (a rotation never fails on this). */
    private val recoveryEnrolled: suspend () -> Boolean = {
        runCatching { server.hasCloudRecovery(username) }.getOrDefault(false)
    },
    /** Re-wrap the NEW root under the EXISTING recovery credential (D-3 —
     *  CloudRecoveryEnrollment.reEscrowAdminRoot). The ceremony needs the
     *  foreground Activity, so the host composable injects the production
     *  wiring (mirroring how RecoveryScreen wires enroll()). */
    private val reEscrow: suspend (passphrase: String, newSeed: ByteArray) -> Unit = { _, _ ->
        error("re-escrow isn't wired on this surface")
    },
) : ViewModel() {

    private val _phase = MutableStateFlow<RotateAdminRootPhase>(RotateAdminRootPhase.Idle)
    val phase: StateFlow<RotateAdminRootPhase> = _phase.asStateFlow()

    /** The NEW seed, held ONLY between a successful rotate and the re-escrow
     *  resolution (update or skip). Never surfaced on a Phase. */
    private var pendingNewSeed: ByteArray? = null

    /** The UI greys the control out when this device holds no admin root. */
    fun canRotate(): Boolean = hasAdminRoot()

    /**
     * Rotate the admin master root: load the OLD root (biometric), mint a NEW
     * random root, sign `old→new`, POST the proof to `.com`, then re-store the
     * NEW root. With cloud recovery enrolled, park in DoneNeedsRecoveryUpdate
     * so the owner can re-escrow the NEW root (the rotation itself is already
     * committed either way).
     */
    suspend fun rotate() {
        if (!hasAdminRoot()) {
            _phase.value = RotateAdminRootPhase.Failed(
                "This device isn't an admin, so it can't rotate the admin key.",
            )
            return
        }
        _phase.value = RotateAdminRootPhase.Rotating
        try {
            val oldSigner = loadOldSigner() // biometric-gated read of the OLD root
            val old = (oldPubHex() ?: error("no admin root on this device")).lowercase()
            val newSeed = mintSeed()
            require(newSeed.size == 32) { "new admin root seed must be 32 bytes" }
            val newPub = HexUtil.encode(Ed25519Sign.KeyPair.newKeyPairFromSeed(newSeed).publicKey)
            val issuedAt = now()

            // OLD signs OLD→NEW — byte-identical canonical bytes to the TS spine.
            val rotation = AdminRootRotation(
                username = username,
                oldAdminRootPub = old,
                newAdminRootPub = newPub,
                issuedAt = issuedAt,
            )
            val signatureHex = HexUtil.encode(AdminRootRotationClaim.sign(rotation, oldSigner))

            server.rotateAdminRoot(
                username = username,
                req = AdminRootRotationRequest(
                    rotation = AdminRootRotationRequest.Rotation(
                        username = username,
                        oldAdminRootPub = old,
                        newAdminRootPub = newPub,
                        issuedAt = issuedAt,
                    ),
                    signatureHex = signatureHex,
                ),
            )

            // Re-store the NEW root device-local (the OLD root is now dead).
            storeNewRoot(newSeed)

            // The rotation is committed from here on — nothing below may fail it.
            val enrolled = runCatching { recoveryEnrolled() }.getOrDefault(false)
            if (enrolled) {
                pendingNewSeed = newSeed
                _phase.value = RotateAdminRootPhase.DoneNeedsRecoveryUpdate(newPub)
            } else {
                _phase.value = RotateAdminRootPhase.Done(newPub)
            }
        } catch (t: Throwable) {
            _phase.value = RotateAdminRootPhase.Failed(
                t.message ?: "Couldn't rotate the admin key. Try again.",
            )
        }
    }

    /**
     * Re-wrap the NEW root under the existing recovery credential. Success →
     * Done; failure keeps DoneNeedsRecoveryUpdate with [RotateAdminRootPhase
     * .DoneNeedsRecoveryUpdate.errorMessage] set so the owner can retry.
     */
    suspend fun updateRecoveryBackup(passphrase: String) {
        val current = _phase.value as? RotateAdminRootPhase.DoneNeedsRecoveryUpdate ?: return
        val seed = pendingNewSeed
        if (seed == null) {
            _phase.value = RotateAdminRootPhase.Done(current.newAdminRootPubHex)
            return
        }
        _phase.value = current.copy(updating = true, errorMessage = null)
        try {
            reEscrow(passphrase, seed)
            pendingNewSeed = null
            _phase.value = RotateAdminRootPhase.Done(current.newAdminRootPubHex)
        } catch (t: Throwable) {
            _phase.value = current.copy(
                updating = false,
                errorMessage = t.message ?: "Couldn't update the recovery backup. Try again.",
            )
        }
    }

    /** Leave the recovery envelope holding the DEAD old root (owner's call —
     *  the UI shows a standing caution). */
    fun skipRecoveryUpdate() {
        val current = _phase.value as? RotateAdminRootPhase.DoneNeedsRecoveryUpdate ?: return
        pendingNewSeed = null
        _phase.value = RotateAdminRootPhase.Done(current.newAdminRootPubHex)
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/RotateAdminRootViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

/// Slice D §5 — the owner-facing "Rotate admin key" action (Account security /
/// a "my device may be compromised" flow).
///
/// Rotation is the compromise remedy AND the revoke-every-other-admin remedy.
/// It:
///   1. loads the CURRENT admin master root (biometric),
///   2. mints a fresh RANDOM new admin master root,
///   3. signs the spine `admin-root-rotation/v1` envelope with the OLD root
///      (`old → new`, the box verifies it against its pinned OLD root — `.com`
///      is never a trust anchor),
///   4. POSTs it to `.com` `POST /api/users/:username/admin-root-rotation`,
///   5. only on success re-seals the NEW root device-local,
///   6. then — if recovery is enrolled — asks the user for the interactive
///      re-escrow step (§5.3 / D-3): the recovery envelope still wraps the
///      OLD root, so a later credential recovery would restore a dead key
///      until the user runs `updateRecoveryBackup(passphrase:)` (recovery
///      passphrase + WebAuthn PRF against the EXISTING credential; see
///      `AdminRootReEscrow`).
///
/// ⚠️ Rotation EXCLUDES every OTHER admin device: they still hold the OLD root,
/// so once each box re-pins to the new root their old-root-signed orders stop
/// verifying. That IS the "revoke an admin" / "a device may be stolen"
/// semantic — there is no separate per-device admin-revoke; you rotate and the
/// non-recovering admins fall out. A device promoted via a master-root SEAL
/// (bare-root admin) can only be removed this way (it holds a key, not a
/// grant), which is exactly the accepted trade-off in the spec (§4.3).
///
/// Order matters: sign + POST FIRST, and re-seal the NEW root locally only
/// AFTER `.com` accepts it. If the POST fails we must NOT have already
/// replaced this device's sealed root — that would strand it (old root gone,
/// new root unrecorded). So the local re-seal is the LAST mutating step, and
/// the recovery re-escrow comes strictly AFTER it: by then the rotation is
/// published + sealed, so a re-escrow failure (or skip) can never fail the
/// rotation itself — it only leaves the recovery backup pointing at the old
/// root, which the user can fix later by re-running recovery setup.
@Observable
@MainActor
public final class RotateAdminRootViewModel {

    public enum Phase: Equatable, Sendable {
        case idle
        case rotating
        /// Rotation accepted + the new root re-sealed. `newAdminRootPubHex` is
        /// the account's new authority anchor (lowercased hex).
        case rotated(newAdminRootPubHex: String)
        /// Rotation accepted + sealed, but recovery is enrolled and its
        /// envelope still wraps the OLD root — the UI renders the inline
        /// re-escrow step (passphrase → `updateRecoveryBackup`, or
        /// `skipRecoveryUpdate`). The rotation itself is DONE in this state.
        case rotatedNeedsRecoveryUpdate(newAdminRootPubHex: String)
        case failed(String)
    }

    public private(set) var phase: Phase = .idle
    /// True while `updateRecoveryBackup` is in flight (drives the button's
    /// progress state).
    public private(set) var isUpdatingRecoveryBackup = false
    /// Inline, retryable error from the last `updateRecoveryBackup` attempt.
    public private(set) var recoveryUpdateError: String?
    /// Set by `skipRecoveryUpdate` so the UI can warn that the recovery
    /// backup still wraps the OLD admin root.
    public private(set) var didSkipRecoveryUpdate = false

    private let server: any FlagshipServerClient
    private let username: @MainActor () -> String?

    /// Seam: does THIS device hold the admin master root? Only such a device
    /// can rotate (it must sign `old → new` with the OLD root). Defaults to
    /// `Keystore.hasAdminRoot`.
    private let hasAdminRoot: @MainActor () -> Bool
    /// Seam: unseal the CURRENT admin master root (biometric). Defaults to
    /// `Keystore.adminRootKey`.
    private let loadOldAdminRoot: @MainActor (String) async throws -> Curve25519.Signing.PrivateKey
    /// Seam: seal the freshly-minted NEW admin root seed device-local.
    /// Defaults to `Keystore.importAdminRoot`.
    private let sealNewAdminRoot: @MainActor (Data) async throws -> Void
    /// Seam: is WebAuthn-PRF cloud recovery enrolled for this account? Decides
    /// whether a successful rotation lands in `.rotatedNeedsRecoveryUpdate`
    /// (the envelope wraps the OLD root and needs the interactive re-escrow)
    /// or straight in `.rotated`. A throw is treated as "not enrolled" — the
    /// check is best-effort and must NEVER fail an already-published rotation.
    /// Defaults to `server.hasCloudRecovery`.
    private let recoveryEnrolled: @MainActor () async throws -> Bool
    /// Seam: the interactive re-escrow of the CURRENT admin root under the
    /// existing recovery credential (§5.3 / D-3), taking the user's recovery
    /// passphrase. Defaults to `AdminRootReEscrow` built on the production
    /// WebAuthn provider — the passphrase + PRF assert EMIT the wrap key, so
    /// the step is user-driven, not automatic.
    private let reEscrow: @MainActor (String) async throws -> Void

    private struct NoAccountError: LocalizedError {
        var errorDescription: String? { "No active account on this device." }
    }

    public init(
        server: any FlagshipServerClient,
        username: @escaping @MainActor () -> String?,
        hasAdminRoot: @escaping @MainActor () -> Bool = { Keystore.hasAdminRoot },
        loadOldAdminRoot: @escaping @MainActor (String) async throws -> Curve25519.Signing.PrivateKey = { reason in
            try await Keystore.adminRootKey(reason: reason)
        },
        sealNewAdminRoot: @escaping @MainActor (Data) async throws -> Void = { seed in
            _ = try await Keystore.importAdminRoot(seed: seed, reason: "Save your new admin key")
        },
        recoveryEnrolled: (@MainActor () async throws -> Bool)? = nil,
        reEscrow: (@MainActor (String) async throws -> Void)? = nil
    ) {
        self.server = server
        self.username = username
        self.hasAdminRoot = hasAdminRoot
        self.loadOldAdminRoot = loadOldAdminRoot
        self.sealNewAdminRoot = sealNewAdminRoot
        self.recoveryEnrolled = recoveryEnrolled ?? {
            guard let user = username(), !user.isEmpty else { return false }
            return try await server.hasCloudRecovery(username: user)
        }
        self.reEscrow = reEscrow ?? { passphrase in
            guard let user = username(), !user.isEmpty else { throw NoAccountError() }
            try await AdminRootReEscrow(
                client: server,
                webAuthn: PlatformWebAuthnProvider()
            ).run(username: user, passphrase: passphrase)
        }
    }

    /// True iff the control should be enabled — only a device that holds the
    /// master root can drive a rotation (§8.2).
    public var canRotate: Bool { hasAdminRoot() }

    /// Run the rotation. Idempotent-safe to retry on failure (nothing local is
    /// mutated until `.com` accepts).
    public func rotate() async {
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        guard hasAdminRoot() else {
            phase = .failed("This device isn't an admin, so it can't rotate the admin key.")
            return
        }
        phase = .rotating

        // 1 — the OLD root (the box's pinned anchor + the rotation signer).
        let oldRoot: Curve25519.Signing.PrivateKey
        do {
            oldRoot = try await loadOldAdminRoot("Rotate your admin key")
        } catch {
            phase = .failed("Couldn't access your admin key: \(error.localizedDescription)")
            return
        }
        let oldPubHex = HexUtil.encode(oldRoot.publicKey.rawRepresentation)

        // 2 — a fresh RANDOM new root (NOT derived — the authority root never is).
        let newRoot = Curve25519.Signing.PrivateKey()
        let newPubHex = HexUtil.encode(newRoot.publicKey.rawRepresentation)

        // 3 — sign the spine `old → new` proof with the OLD root.
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let rotation = AdminRootRotation(
            username: user,
            oldAdminRootPubHex: oldPubHex,
            newAdminRootPubHex: newPubHex,
            issuedAt: issuedAt
        )
        let signatureHex: String
        do {
            signatureHex = HexUtil.encode(try rotation.sign(withOldAdminRoot: oldRoot))
        } catch {
            phase = .failed("Couldn't sign the rotation: \(error.localizedDescription)")
            return
        }

        // 4 — publish. On any failure we STOP here: the local sealed root is
        // still the OLD one, so this device remains a working admin.
        do {
            _ = try await server.postAdminRootRotation(
                username: user,
                body: AdminRootRotationRequest(
                    rotation: .init(
                        username: user,
                        oldAdminRootPub: oldPubHex,
                        newAdminRootPub: newPubHex,
                        issuedAt: issuedAt
                    ),
                    signatureHex: signatureHex
                )
            )
        } catch {
            phase = .failed("Couldn't rotate your admin key: \(error.localizedDescription)")
            return
        }

        // 5 — .com accepted; NOW replace the local root. A re-seal failure
        // here leaves the account rotated at `.com`/boxes but this device
        // still holding the OLD root — surface it so the user can recover
        // onto the new root.
        do {
            try await sealNewAdminRoot(newRoot.rawRepresentation)
        } catch {
            phase = .failed("Your admin key was rotated, but saving it on this device failed. Recover on this device to finish. (\(error.localizedDescription))")
            return
        }

        // 6 — the rotation is DONE (published + sealed); everything from here
        // is best-effort and must not undo that. If recovery is enrolled, the
        // envelope still wraps the OLD root — hand the UI the interactive
        // re-escrow step. An enrollment-check failure ⇒ treat as not enrolled
        // (the user can always re-run recovery setup later).
        didSkipRecoveryUpdate = false
        recoveryUpdateError = nil
        let enrolled = (try? await recoveryEnrolled()) ?? false
        phase = enrolled
            ? .rotatedNeedsRecoveryUpdate(newAdminRootPubHex: newPubHex)
            : .rotated(newAdminRootPubHex: newPubHex)
    }

    /// Run the interactive re-escrow (§5.3 / D-3) with the user's recovery
    /// passphrase. Success completes the flow (`.rotated`); failure keeps the
    /// step on screen with an inline, retryable error — the rotation itself is
    /// already done either way.
    public func updateRecoveryBackup(passphrase: String) async {
        guard case .rotatedNeedsRecoveryUpdate(let newPubHex) = phase else { return }
        recoveryUpdateError = nil
        isUpdatingRecoveryBackup = true
        defer { isUpdatingRecoveryBackup = false }
        do {
            try await reEscrow(passphrase)
            phase = .rotated(newAdminRootPubHex: newPubHex)
        } catch {
            recoveryUpdateError = error.localizedDescription
        }
    }

    /// Decline the re-escrow step. The recovery envelope keeps wrapping the
    /// OLD admin root (a later credential recovery restores a dead key) — the
    /// UI warns via `didSkipRecoveryUpdate`; re-running recovery setup fixes it.
    public func skipRecoveryUpdate() {
        guard case .rotatedNeedsRecoveryUpdate(let newPubHex) = phase else { return }
        didSkipRecoveryUpdate = true
        recoveryUpdateError = nil
        phase = .rotated(newAdminRootPubHex: newPubHex)
    }

    public func reset() {
        phase = .idle
        recoveryUpdateError = nil
        didSkipRecoveryUpdate = false
    }
}
